Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,354 @@
+
+It needs to be possible to call this function simultaneously from multiple threads at the same time.
+It also needs to be possible to call this function while there is an active host context created by `hostfxr_initialize_for_app` and running inside the `hostfxr_run_app`.
+
+The function returns specific return code for the first initialized host context, and a different one for any subsequent one. Both return codes are considered "success". If there already was initialized host context in the process then the returned host context has these limitations:
+* It won't allow setting runtime properties.
+* The initialization will compare the runtime properties from the `.runtimeconfig.json` specified in the `runtime_config_path` with those already set to the runtime in the process
+  * If all properties from the new runtime config are already set and have the exact same values (case sensitive string comparison), the initialization succeeds with no additional consequences. (Note that this is the most typical case where the runtime config have no properties in it.)
+  * If there are either new properties which are not set in the runtime or ones which have different values, the initialization will return a special return code - a "warning". It's not a full on failure as initialized context will be returned.
+  * In both cases only the properties specified by the new runtime config will be reported on the host context. This is to allow the native host to decide in the "warning" case if it's OK to let the component run or not.
+  * In both cases the returned host context can still be used to get a runtime delegate, the properties from the new runtime config will be ignored (as there's no way to modify those in the runtime).
+
+
+### Inspect and modify host context
+
+#### Runtime properties
+These functions allow the native host to inspect and modify runtime properties.
+* If the `host_context_handle` represents the first initialized context in the process, these functions expose all properties from runtime configurations as well as those computed by the hosting layer components. These functions will allow modification of the properties via `hostfxr_set_runtime_property`. 
+* If the `host_context_handle` represents any other context (so not the first one), these functions expose only properties from runtime configuration. These functions won't allow modification of the properties.
+
+It is possible to access runtime properties of the first initialized context in the process at any time (for reading only), by specifying `nullptr` as the `host_context_handle`.
+
+``` C
+int hostfxr_get_runtime_property(
+    const hostfxr_handle host_context_handle,
+    const char_t * name,
+    char_t * buffer,
+    size_t buffer_size,
```

Comment:
It is a common Win32 idion to use a single in/out to indicate count (i.e. it is read for max and then updated to return written).

Reply:
Alternative the internal property value could just be returned thus avoiding any allocation: (i.e. `const char_t **current_value`). There is the issue of a bad actor casting away `const` and creating havoc, but if a caller can access the hosting API layer that doesn't seem to be much of an increase in attack vector.